src/group_theory/free_abelian_group_finsupp.lean

```diff
@@ -94,6 +97,51 @@ def equiv_finsupp : free_abelian_group X ≃+ (X →₀ ℤ) :=
   right_inv := to_finsupp_to_free_abelian_group,
   map_add' := to_finsupp.map_add }
 
+ /-- we now obtain that `A` is a basis for `free_abelian_group X`. -/
+noncomputable def free_abelian_group.basis (α : Type*) :
+  basis α ℤ (free_abelian_group α) :=
+⟨add_equiv.to_int_linear_equiv (free_abelian_group.equiv_finsupp α)⟩
+
+/-- Isomorphic free ablian groups (as modules) have equivalent bases. -/
+def equiv.of_free_abelian_group_linear_equiv {α β : Type*}
+  (e : free_abelian_group α ≃ₗ[ℤ] free_abelian_group β) :
+  α ≃ β :=
+let t : basis α ℤ (free_abelian_group β) := (free_abelian_group.basis α).map e
+  in t.index_equiv $ free_abelian_group.basis _
+
+/-- Isomorphic free abelian groups (as additive groups) have equivalent bases. -/
+def equiv.of_free_abelian_group_equiv {α β : Type*}
+  (e : free_abelian_group α ≃+ free_abelian_group β) :
+  α ≃ β :=
+equiv.of_free_abelian_group_linear_equiv (add_equiv.to_int_linear_equiv e)
+
+/-- Isomorphic free groups have equivalent bases. -/
+def equiv.of_free_group_equiv {α β : Type*}
+  (e : free_group α ≃* free_group β) :
+  α ≃ β :=
+equiv.of_free_abelian_group_equiv $
+  mul_equiv.to_additive (abelianization.mul_equiv.abelianization_congr e)
+
+open is_free_group
+/-- Isomorphic free groups have equivalent bases (`is_free_group` variant`). -/
+def equiv.of_is_free_group_equiv {G H : Type*}
+  [group G] [group H] [is_free_group G] [is_free_group H]
+  (e : G ≃* H) :
+  generators G ≃ generators H :=
+equiv.of_free_group_equiv $
+  mul_equiv.trans ((to_free_group G).symm) $
+  mul_equiv.trans e $
+  to_free_group H
```

Comment:
Done. The proof using `calc` did not work, but writing out the proof term using `mul_equiv.trans` did and I would be curious why. Something with universe levels maybe? 
Here is what I tried
```
calc free_group (generators G) ≃* G : (to_free_group G).symm
  ... ≃* H : e
  ... ≃* free_group (generators H) : to_free_group H```
```
and the error message is
```
type mismatch at application
  mul_equiv.trans
term
  free_group (generators G)
has type
  Type ? : Type (?+1)
but is expected to have type
  has_mul ?m_1 : Type ?
```

Reply:
Maybe the implicit parameters of `≃*` make it unsuitable for use in `calc`?

Reply:
I am not an expert in `calc`, but `mul_equiv` is a `structure`, and the doc only speaks about relations. Does it work for an `equiv`?

Reply:
I guess it would work if you define a relation `is_equiv`, but the actual `equiv` contains more data.

Reply:
Thanks. I’ll just leave it as it is now.